### PR TITLE
Implement CA1303 (DoNotPassLiteralsAsLocalizedParameters) using Dataf…

### LINKED ISSUE #411: Port FxCop rule CA1303: DoNotPassLiteralsAsLocalizedParameters
**Title:** Do not pass literals as localized parameters

**Description:**

An externally visible method passes a string literal as a parameter to a constructor or method in the .NET Framework class library, and that string should be localizable.

**Dependency:** Dataflow

**Notes:**

### PR DESCRIPTION
**Pull request by mavasani**
…low StringContentAnalysis

_A method passes a string literal as a parameter to a constructor or method in the .NET Framework class library and that string should be localizable._

Rule documentation: https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1303-do-not-pass-literals-as-localized-parameters

Fixes #411

### CODE REVIEW

**Comment by mavasani** [comment]
Ping.. this is a trivial rule and the PR does not change or add any flow analysis, so needs reviewing just the core analyzer implementation and unit tests.

File: src/Microsoft.CodeQuality.Analyzers.Exp/Core/Globalization/DoNotPassLiteralsAsLocalizedParameters.cs [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+            context.RegisterCompilationStartAction(compilationContext =>
+            {
+                INamedTypeSymbol localizableStateAttributeSymbol = WellKnownTypes.LocalizableAttribute(compilationContext.Compilation);
+                INamedTypeSymbol conditionalAttributeSymbol = WellKnownTypes.ConditionalAttribute(compilationContext.Compilation);
+                INamedTypeSymbol systemConsoleSymbol = WellKnownTypes.Console(compilationContext.Compilation);
+                ImmutableHashSet<INamedTypeSymbol> typesToIgnore = GetTypesToIgnore(compilationContext.Compilation);
+
+                compilationContext.RegisterOperationBlockStartAction(operationBlockStartContext =>
+                {
+                    if (!(operationBlockStartContext.OwningSymbol is IMethodSymbol containingMethod))
```
Comment (333fred) [inline src/Microsoft.CodeQuality.Analyzers.Exp/Core/Globalization/DoNotPassLiteralsAsLocalizedParameters.cs]: Is this FxCop compat? This doesn't seem like it's correct, we create LocalizableResourceStrings as fields in this very class.

Comment (mavasani) [reply src/Microsoft.CodeQuality.Analyzers.Exp/Core/Globalization/DoNotPassLiteralsAsLocalizedParameters.cs]: Yes, this is not FxCop compat, but restriction of our initial flow analysis implementation that doesn't work on initializers. Let me file a tracking issue to remove this workaround.

**Comment by 333fred** [comment]
> // Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.

I don't see any tests for readonly properties initialized to constants. Is this not a supported scenario? I also don't see any for constant fields.

---
Refers to: src/Microsoft.CodeQuality.Analyzers.Exp/UnitTests/Globalization/DoNotPassLiteralsAsLocalizedParametersTests.cs:1 in cdb5cd8. [](commit_id = cdb5cd8d9041f17f04c67045552ac6a8e0147812, deletion_comment = False)

**Comment by mavasani** [comment]
Ping @333fred @ivanbasov, can you please review?

We would like to include this rule in the experimental package we hand out to internal customers by EOD today.

**Comment by mavasani** [comment]
Thanks @jinujoseph. I will merge and have a follow up PR for feedback from prior PRs.

**Pull request merged**